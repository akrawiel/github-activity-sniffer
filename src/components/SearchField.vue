<template lang="pug">
.relative
  input.search-field(:class="{ error: hasError }" type="search" placeholder="Enter GitHub organization's name here" autoFocus @input="search")
  search-spinner(:isFetching="isFetching")
</template>

<script>
import debounce from "lodash.debounce";
import SearchSpinner from "@/components/SearchSpinner";
import { getOrganizationByName } from "@/api/github-api";

export default {
  components: {
    SearchSpinner
  },
  data() {
    return {
      isFetching: false,
      errorMessage: "",
      showErrorMessage: true,
      searchDebounced: debounce(async function({ searchValue }) {
        await this.transformOrganizationResponse({
          promise: () =>
            getOrganizationByName({ organizationName: searchValue })
        });

        this.isFetching = false;
      }, 500)
    };
  },
  props: {
    transformOrganizationResponse: Function,
    hasError: Boolean
  },
  methods: {
    search(event) {
      const searchValue = event.target.value;

      this.$emit("clear");
      if (searchValue) {
        this.isFetching = true;
        this.searchDebounced({ searchValue });
      } else {
        this.searchDebounced.cancel();
        this.isFetching = false;
        this.$emit("error", "Enter at least 1 character in the field above");
      }
    }
  }
};
</script>

<style lang="postcss" scoped>
.search-field {
  @apply border border-gray-400 border-solid p-3 pr-4 rounded shadow-lg w-full;
  transition: border-color 0.2s ease-out;

  &:focus {
    @apply border-gray-600;
  }

  &.error {
    @apply border-red-500;
  }
}
</style>
