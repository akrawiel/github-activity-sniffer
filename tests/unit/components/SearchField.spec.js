import { shallowMount } from "@vue/test-utils";
import SearchField from "@/components/SearchField.vue";

const dummyValidEvent = {
  target: { value: "test" }
};

const dummyInvalidEvent = {
  target: { value: "" }
};

describe("SearchField.vue", () => {
  it("renders when all data is passed", () => {
    const propsData = {
      transformOrganizationResponse: jest.fn(),
      hasError: false
    };

    const wrapper = shallowMount(SearchField, {
      propsData
    });

    expect(wrapper.element).toMatchSnapshot();

    const searchDebouncedSpy = jest.fn(() => {
      wrapper.vm.transformOrganizationResponse();
    });
    searchDebouncedSpy.cancel = jest.fn();

    wrapper.setData({ searchDebounced: searchDebouncedSpy });

    // called with valid event
    wrapper.vm.search(dummyValidEvent);

    expect(wrapper.emitted("clear")).toHaveLength(1);
    expect(wrapper.vm.isFetching).toBeTruthy();
    expect(searchDebouncedSpy).toHaveBeenCalledWith({
      searchValue: dummyValidEvent.target.value
    });
    expect(propsData.transformOrganizationResponse).toHaveBeenCalled();

    searchDebouncedSpy.mockReset();

    // called with invalid event
    wrapper.vm.search(dummyInvalidEvent);

    expect(wrapper.emitted("clear")).toHaveLength(2);
    expect(wrapper.vm.isFetching).toBeFalsy();
    expect(searchDebouncedSpy).not.toHaveBeenCalled();
    expect(searchDebouncedSpy.cancel).toHaveBeenCalled();
    expect(wrapper.emitted("error")).toMatchObject([
      ["Enter at least 1 character in the field above"]
    ]);
  });
});
