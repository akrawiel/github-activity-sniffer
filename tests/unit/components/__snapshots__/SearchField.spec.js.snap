// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchField.vue renders when all data is passed 1`] = `
<div
  class="relative"
>
  <input
    autofocus="autoFocus"
    class="search-field"
    placeholder="Enter GitHub organization's name here"
    type="search"
  />
  <search-spinner-stub />
</div>
`;
